### PR TITLE
Add a registry cache

### PR DESCRIPTION
**Pull request by paulbellamy**
Not a massive time-save:
```
$ time fluxctl list-images -s monitoring/kubediff
SERVICE              CONTAINER  IMAGE                CREATED
monitoring/kubediff  git-sync   tomwilkie/git-sync
                                '-> f6165715ce9d     02 Jun 16 12:39 UTC
                                    latest           02 Jun 16 12:39 UTC
                     kubediff   weaveworks/kubediff
                                |   latest           16 Nov 16 14:57 UTC
                                |   master-d658306   16 Nov 16 14:57 UTC
                                |   master-8ac1b99   06 Nov 16 17:34 UTC
                                |   master-15b51e6   18 Oct 16 12:58 UTC
                                |   master-d82bdad   18 Oct 16 11:55 UTC
                                |   master-af2af60   18 Oct 16 11:03 UTC
                                |   master-f332982   18 Oct 16 10:53 UTC
                                |   master-139e1ee   05 Oct 16 01:19 UTC
fluxctl list-images -s monitoring/kubediff  0.01s user 0.01s system 0% cpu 3.084 total

# Next time, cached...
$ time fluxctl list-images -s monitoring/kubediff
SERVICE              CONTAINER  IMAGE                CREATED
monitoring/kubediff  git-sync   tomwilkie/git-sync
                                '-> f6165715ce9d     02 Jun 16 12:39 UTC
                                    latest           02 Jun 16 12:39 UTC
                     kubediff   weaveworks/kubediff
                                |   latest           16 Nov 16 14:57 UTC
                                |   master-d658306   16 Nov 16 14:57 UTC
                                |   master-8ac1b99   06 Nov 16 17:34 UTC
                                |   master-15b51e6   18 Oct 16 12:58 UTC
                                |   master-d82bdad   18 Oct 16 11:55 UTC
                                |   master-af2af60   18 Oct 16 11:03 UTC
                                |   master-f332982   18 Oct 16 10:53 UTC
                                |   master-139e1ee   05 Oct 16 01:19 UTC
fluxctl list-images -s monitoring/kubediff  0.01s user 0.01s system 0% cpu 1.872 total
```
but should reduce quite a few round-trips to the registry.

Note, I've just picked 20 minutes as the time to cache entries for. Completely out of the air, no data basis for that. but it *should* lead to a 1/20th of the requests.

Note: to deploy, need to add memcached first, then add the flags. If memcached-host flag is not set, no behaviour change.

### CODE REVIEW

File: instance/multi.go [hunk 1]
```diff
@@ -47,6 +50,8 @@ func (m *MultitenantInstancer) Get(instanceID flux.InstanceID) (*Instance, error
 		creds,
 		log.NewContext(instanceLogger).With("component", "registry"),
 		m.RegistryMetrics.WithInstanceID(instanceID),
+		m.MemcacheClient,
```
Comment (philwinder) [inline instance/multi.go]: Memcache client should be a middleware for the docker client. I can refactor this in a subsequent commit with my other registry changes.

File: registry/registry.go [hunk 2]
```diff
@@ -46,20 +47,29 @@ type Client interface {
 
 // client is a handle to a registry.
 type client struct {
-	Credentials Credentials
-	Logger      log.Logger
-	Metrics     Metrics
+	Credentials    Credentials
+	Logger         log.Logger
```
Comment (philwinder) [inline registry/registry.go]: Yeah, if middleware, wouldn't need to edit all this.

File: registry/registry.go [hunk 3]
```diff
@@ -126,16 +136,23 @@ func (c *client) GetRepository(repository string) (_ []flux.ImageDescription, er
...
 
-	client := &dockerregistry.Registry{
-		URL: httphost,
-		Client: &http.Client{
-			Transport: roundtripperFunc(func(r *http.Request) (*http.Response, error) {
-				return transport.RoundTrip(r.WithContext(ctx))
-			}),
-			Jar:     jar,
-			Timeout: 10 * time.Second,
+	var client backend = herokuWrapper{
```
Comment (philwinder) [inline registry/registry.go]: This looks new. Might be tempted to do this in separate PR.

Comment (paulbellamy) [inline registry/registry.go]: maybe just a separate commit? but the motivation is coupled to serializing data to send to memcache. so, I'm not sure.
